src/server/transaction.cc

```diff
@@ -323,6 +329,32 @@ OpStatus Transaction::InitByArgs(DbIndex index, CmdArgList args) {
   return OpStatus::OK;
 }
 
+void Transaction::PrepareSquashedMultiHop(const CommandId* cid, CmdArgList keys) {
+  MultiSwitchCmd(cid);
+
+  InitBase(db_index_, keys);
+  InitByKeys(KeyIndex::Range(0, keys.size()));
+}
+
+void Transaction::PrepareSquashedMultiHop(const CommandId* cid,
+                                          std::function<bool(ShardId)> enabled) {
+  CHECK(multi_->mode == GLOBAL || multi_->mode == LOCK_AHEAD);
+
+  MultiSwitchCmd(cid);
+  InitBase(db_index_, {});
+
+  DCHECK_EQ(shard_data_.size(), shard_set->size());
+  for (unsigned i = 0; i < shard_data_.size(); i++) {
+    if (enabled(i)) {
+      shard_data_[i].local_mask |= ACTIVE;
+      unique_shard_cnt_++;
+      unique_shard_id_ = i;
+    } else {
+      shard_data_[i].local_mask &= ~ACTIVE;
+    }
+  }
```

Comment:
I feel really uncomfortable having duplicated initialization code - it makes tracking state a lot more complicated. I think I'll re-use `InitShardData` for this even though it requires this temporary per-shard-cache stuff